Add firewall rule to accept traffic from the netns

Belphemur opened the pull request:
Depending on iptables default rule, if we don't add the acceptance of input, the root network cannot accept return packet from the netns.

Also cleanup an `exec` that isn't needed.

Belphemur committed:
Add Firewall rule to accept traffic from the netns

Belphemur committed:
No need to use exec for IP command

Belphemur committed:
Fixes start instead of stop

JoelStienlet commented:
Hello,
I'm not a maintainer but I've seen the patch in my notifications, and there is something I don't get about the firewall rule change suggested, so I wished to discuss it: when looking on this chart:
https://stuffphilwrites.com/2014/09/iptables-processing-flowchart/
we see that the INPUT chain and the FORWARD chain handle completely different traffic.
Why should the namespace get some access to the root network by default? I know netns has not been designed for security, but this new rule changes the behaviour of isolation between the two networks, so it could expose "root nework" services that weren't previously exposed to the namespace network, maybe create security problems?

Belphemur commented:
Hello @JoelStienlet,

Those are very valid points that I didn't take into account when making the patch.
I guess, I could easily put this "feature" as a configurable option in the Environment file / Script file.

The reason why I added those iptables rules is to let the root namespace have access to the application running into the netns that is bound on the veth.

Actually, the root netns has already access to the created ns, but never get an answer back to its requests.

In my case, the application running in the netns has an RPC api that I want accessible from the root netns, while the rest of the traffic the app does, I want it to be managed by the created netns.

JoelStienlet commented:
I think it's a good idea to add it as a configurable feature (and disable it by default so that updates won't change the behaviour of existing installations). Looks useful indeed, as the requirement to access the netns from the root is certainly pretty common.
What do you think about adding connection tracking to restrict the access a bit more?
iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

Jamesits commented:
@Belphemur I've merged your changes on the `ip -n` individually. I was not aware of this usage. Thank you. 

My thoughts on the other new feature:

When I wrote these scripts in the very beginning, I expected them to function as some kind of VRF-lite thing (if you are familiar with Cisco terms), i.e. different routing tables on one network stack and we can put different programs into different routing tables while sharing the physical network interfaces and other system resources. At that time the security on the INPUT chain is out of my consideration since CoPP (Control Plane Policing, i.e. rules to control if some service on the localhost is accessible by someone) has no relationship with VRF-lite from a network engineering guy's intuition. 

Also, I'm strongly opposed to adding rules on the filter table automatically by default. My major considerations are:

* You don't know where this rule will be placed
* You don't know how it will affect other rules or be affected by other rules
* Every distro has its defaults on iptables filter rules (e.g. Debian defaults to ACCEPT, CentOS defaults to DROP), and it is generally not good to break people's assumptions

The use of conntrack module is another thing that needs to be evaluated case by case. There are many cases where you need to not using it or even unload conntrack-related modules from your kernel in order to get some performance. 

Although being said, it is still possible you need to run certain programs after `netns-*` has been set up or shut down. This project doesn't come with up/down/pre-up/pre-down script support since systemd natively support this via its drop-in feature. (I'd admit this is not a very Linux way comparing to a trigger script, but this project is heavily dependent on systemd, let's just embrace a more systemd-native approach.)

FYI, documentation on systemd drop-in files:
https://wiki.archlinux.org/index.php/systemd#Drop-in_files

Jamesits commented:
@Belphemur I think it would be nice if you make this configurable (and not enabled by default) and I would be happy to merge this.

Belphemur committed:
Make Nat Access configurable (#1)

* Add NAT_ACCEPT_TRAFFIC

to be able to configure if we want to accept traffic or not from the child namespace.

* Add up and down for nat setup

To be able to use IPTABLES for configure traffic.

* Replace the iptable to a call to the script

* Split net access in own method

* Fix script to be run in the root namespace

* Don't mixup start and stop

* Add missing selector

* Add missing spaces

* Fix iptables parameters

Belphemur commented:
@Jamesits & @JoelStienlet  As discussed I made it configurable in the default file.

Also I followed your advise @JoelStienlet and use a restricted version of the iptables rule to only accept related traffic.
